[RingCache] Add `EventDelegateeChange` event emission and cache invalidation (#220)
pokt-network/poktroll: docs/pkg/client/events.md

Reviewer: It looks like your latest diagram changes (https://github.com/pokt-network/poktroll/pull/220/commits/18f953c6aaed3840043999903ab9e3a8e536eafc?short_path=877a552#diff-877a552200311af4985616701be119cf6adc302755c4421136aabd6b7b847003) got reverted by the merge with main (e53078bf723a231c5e45b81a3d4cc7973857cde2).

Reviewer: I've added a [section in a "Documentation in progress" page](https://www.notion.so/buildwithgrove/Documentation-in-progress-a2c3bd75c7f14606b2c8e9603d4f907f?pvs=4#b8e57f6bf84b44159cb8bce6ab39469a) that I started on notion.

1. We can use this as a (more) collaborative environment than mermaid.live alone. At the very least, it's better than copy/pasting through discord. :sweat_smile: 
2. I've made some changes that I would've added as a suggestion comment in GH if  the changes hadn't been reverted.

Reviewer: I'm also noticing that the todo's don't render in the docs whereas I was expecting them to. Part of the reason I suggested it was to signal to readers that the legends are missing and to avoid the confusion of applying the wrong legend. Two questions:

1. Is this intentional?
2. Do you think we should prefer this (i.e. rationale, either way)?

Author: I intentionally made them comments but can change to be visible, also unsure what you mean by changes reverted the branch has the latest copy?

  - [Why use the `EventsReplayClient` over directly maintaining an `EventsQueryClient`?](#why-use-the-eventsreplayclient-over-directly-maintaining-an-eventsqueryclient)

<!-- tocstop -->

## Overview

The `events` package provides a client interface to subscribe to chain event
messages. It abstracts the underlying connection mechanisms and offers a clear
and easy-to-use way to get events from the chain. Highlights:

- Offers subscription to chain event messages matching a given query.
- Uses the Gorilla WebSockets package for underlying connection operations.
- Provides a modular structure with interfaces allowing for mock implementations
  and testing.
- Offers considerations for potential improvements and replacements, such as
  integration with the cometbft RPC client.
- Offers a generic client to decode on chain event bytes into the desired event
  type

## Architecture Diagrams

### Components

```mermaid
---
title: Component Diagram Legend
---

flowchart

  a[Component A]
  b[Component B]
  c[Component C]
  d[Component D]

  a --"A uses B via B#MethodName()"--> b
  a =="A returns C from A#MethodName()"==> c

>   b -."A uses D via network IO".-> d

Reviewer: ```suggestion
  b -."B uses D via network IO".-> d
```

```

```mermaid
---
title: EventsQueryClient Components
---

flowchart

  subgraph comet[Cometbft Node]
    subgraph rpc[JSON-RPC]
      sub[subscribe endpoint]
    end
  end

  subgraph eqc[EventsQueryClient]
    q1_eb[EventsBytesObservable]
    q1_conn[Connection]
    q1_dial[Dialer]
  end

  q1_obsvr1[Observer 1]
  q1_obsvr2[Observer 2]

  q1_obsvr1 --"#Subscribe()"--> q1_eb
  q1_obsvr2 --"#Subscribe()"--> q1_eb

  q1_dial =="#DialContext()"==> q1_conn
  q1_eb --"#Receive()"--> q1_conn

  q1_conn -.-> sub
```

```mermaid
---
title: EventsReplayClient Components

> ---

Reviewer: I'm still thinking about this flowchart but the first thing I notice is that it's upside down; take a look at the legend one more time.

Reviewer: I'm noticing that `EventsQueryClient` is described here in terms of inert sub-components. Whereas, in `EventsQueryClient Components`, it's (already) described using sub-components which actually associate with others.

Reviewer: `Publish Events Factory` seems like it should be named `retryPublishEventsFactory`. Additionally, it's a a function represented as a rectangle, but rectangles represent components in the legend (and rest of the diagram). 

In other places I've used a rounded rectangle to represent functions or methods. Regardless of how that "kind" of thing is represented, it should be:

1. Distinct from other "kinds" which are already defined (for other purposes; i.e. component, usage associations, etc.).
2. Used consistently for that "kind", currently `retry.OnError` is depicted as a usage association. It also needs to change and should be represented the same way as `retryPublishEventsFactory` (assuming  both are still represented in the next iteration).
3. Sufficiently described in the legend to be unambiguous (i.e. hard to misinterpret).

Reviewer: FWIW the legend also doesn't currently define rules around encapsulation. That could be taken as an opportunity either to add some and express them in here, or to reconsider representing encapsulation so strictly (e.g. `EventsQueryClient`/`EventsReplayClient`). I don't have a strong opinion here but wanted to share some additional perspective.

Reviewer: I'm wondering if it makes sense for either `retry.OnError` or `retryPublishEventsFactory`  to be in this diagram. This diagram should be about the relationship betwen the components that are necessary for this unit to function. 

I think it makes more sense for `retry.OnError` and `retryPublishEventsFactory` to be central in some other diagram about the retry loop; perhaps a flowchart or state diagram showing the ways in which the client can get disconnected, how it recovers, and ultimately aborts and errors. There should be an analogous diagram for the `EventsQueryClient` as it does a similar thing IIRC. Perhaps they can be represented in the same diagram, but it's okay if not.

Reviewer: I don't think the `#EventBytes` association is accurately represented as network I/O.

Reviewer: > I'm wondering if it makes sense for either `retry.OnError` or `retryPublishEventsFactory` to be in this diagram. This diagram should be about the relationship betwen the components that are necessary for this unit to function.

Similar feedback on `DecoderFunction`. It probably make sense to have a diagram that focuses on data transformation through `EventsReplayClient`, which `DecoderFunction` would be a critical part of. Also, is there a name that's more consistent with the code or have I just not read that part yet?

          q1_conn[Connection]
      end
      subgraph q2[Query 2]
          q2_conn[Connection]
          q2_eb[EventsBytesObservable]
      end
  end

  q1_obsvr1[Query 1 Observer 1]
  q1_obsvr2[Query 1 Observer 2]
  q2_obsvr[Query 2 Observer]

  q1_eb -.-> q1_obsvr1
  q1_eb -.-> q1_obsvr2
  q2_eb -.-> q2_obsvr

  q1_conn -.-> q1_eb
  q2_conn -.-> q2_eb

  sub -.-> q1_conn
  sub -.-> q2_conn
```

## Installation

```bash
go get github.com/pokt-network/poktroll/pkg/client/events
```

## Features

- **Websocket Connection**: Uses the
  [Gorilla WebSockets](https://github.com/gorilla/websocket) for implementing the
  connection interface.
- **Events Subscription**: Subscribe to chain event messages using a simple query
  mechanism.
- **Dialer Interface**: Offers a `Dialer` interface for constructing connections,
  which can be easily mocked for tests.
- **Observable Pattern**: Integrates the observable pattern, making it easier to
  react to chain events.
- **Generic Replay Client**: Offers a generic typed replay client to listen for
  specifc events on chain, and handles reconnection and subscription on error,
  if the `EventsQueryClient` returns an error/is closed.

> ## Usage (`EventsQueryClient`)

Reviewer: I like how you've added structure to make room here. :sunglasses: 

Wdyt about adding links to the godoc? They will eventually be updated after this is merged. In the meantime you can use the `make go_docs` target to double-check the URL path.

for eitherEvent := range observer.Ch() {
    // (see either.Either: https://github.com/pokt-network/poktroll/blob/main/pkg/either/either.go#L3)
    eventBz, err := eitherEvent.ValueOrError()

    // ...
}
```

### Advanced Usage

```go
// Given some custom dialer & connection implementation, e.g.:
var (
    tcpDialer events.Dialer = exampletcp.NewTcpDialerImpl()
    grcpDialer events.Dialer = examplegrpc.NewGrpcDialerImpl()
)

// Both TCP and gRPC use the TCP scheme as gRPC uses TCP for its transport layer.
cometUrl = "tcp://example.com"

// Creating new EventsQueryClients with a custom tcpDialer:
tcpDialerOpt := events.WithDialer(tcpDialer)
tcpEvtClient := events.NewEventsQueryClient(cometUrl, tcpDialerOpt)

// Alternatively, with a custom gRPC dialer:
gcpDialerOpt := events.WithDialer(grcpDialer)
grpcEvtClient := events.NewEventsQueryClient(cometUrl, grpcDialerOpt)

// ... rest follows the same as the basic example.
```

### Configuration

- **WithDialer**: Configure the client to use a custom dialer for connections.

## Usage (`EventsReplayClient`)

### Basic Usage

```go
// Define a query string to provide to the EventsQueryClient

> // See: https://docs.cosmos.network/main/learn/advanced/events#subscribing-to-events
> // And: https://docs.cosmos.network/main/learn/advanced/events#default-events

Reviewer: I had linked to the v0.47 page specifically as some things around events have changed in v0.50 IIRC. I find myself wanting to also suggest adding a TODO to update this (and the analogous comment in `EventsQueryClient`) but I don't think there's a good place for that in the docs. I've started #240 as a way to capture things like this (and I've already added this one).

```suggestion
// See: https://docs.cosmos.network/v0.47/learn/advanced/events#subscribing-to-events
// And: https://docs.cosmos.network/v0.47/learn/advanced/events#default-events
```

const eventQueryString = "tm.event='Tx' AND message.action='eventType'"

// Define the websocket URL the EventsQueryClient will subscribe to

> const cometWebsocketURL = "ws://example.com:36657"

Reviewer: ```suggestion
const cometWebsocketURL = "ws://example.com:36657/websocket"
```

// Define an interface to represent the onchain event
type EventType interface {

>   GetName() string

Reviewer: ```suggestion
  GetName() string // illustrative only; arbitrary interfaces are supported
```

}

// Define the event type that implements the interface
type eventType struct {
  Name string `json:"name"`
}

func (e *eventType) GetName() string { return e.Name }

// Define a decoder function that can take the raw event bytes
// received from the EventsQueryClient and convert them into
// the desired type for the EventsReplayClient
func newEventType(eventBz []byte) EventType {
  eventMsg := new(eventType)
  if err := json.Unmarshal(eventBz, eventMsg); err != nil {
    return nil, err
  }

  // Confirm the event is correct by checking its fields
  if eventMsg.Name == "" {
    return nil, events.ErrEventsUnmarshalEvent.

>       Wrapf("unable to unmarshal eventType: %s", string(eventBz))

Reviewer: Wdyt about something like this?

```suggestion
      Wrapf("with eventType data: %s", string(eventBz))
```

> func newEventType(eventBz []byte) EventType {
>   eventMsg := new(eventType)
>   if err := json.Unmarshal(eventBz, eventMsg); err != nil {
>     return nil, err
>   }
> 
>   // Confirm the event is correct by checking its fields
>   if eventMsg.Name == "" {
>     return nil, events.ErrEventsUnmarshalEvent.
>       Wrapf("unable to unmarshal eventType: %s", string(eventBz))
>   }
> 
>   return eventMsg, nil
> }

Reviewer: Wdyt about making this example use a factory function as that's a pattern that may prove useful in many consumer contexts? We can use a `context.Context` in this example but perhaps there are other and/or more idiomatic factory function parameters.

```suggestion
func eventTypeFactory(ctx context.Context) func ([]byte) EventType {
  return function(eventBz []byte) EventType {  eventMsg := new(eventType)
    logger := polylog.Ctx(ctx)
  
    if err := json.Unmarshal(eventBz, eventMsg); err != nil {
      return nil, err
    }

    // Confirm the event is correct by checking its fields
    if eventMsg.Name == "" {
      return nil, events.ErrEventsUnmarshalEvent.
        Wrapf("unable to unmarshal eventType: %s", string(eventBz))
    }

    return eventMsg, nil
  }
}
```

Reviewer: Perhaps it would also be worth defining a `NewEventFn = func([]byte) EventType` generic (see suggestion above for usage).

// Create the events query client and a depinject config to supply
// it into the EventsReplayClient
evtClient := events.NewEventsQueryClient(cometWebsocketURL)
depConfig := depinject.Supply(evtClient)

// Create a context (this should be cancellable to close the EventsReplayClient)
ctx, cancel := context.WithCancel(context.Background)

// Create a new instance of the EventsReplayClient
client, err := events.NewEventsReplayClient[
  EventType,
  observable.ReplayObservable[EventType],
](
  ctx,
  depConfig,
  cometWebsocketURL,
  eventQueryString,
  newEventType,
)
if err != nil {
  return nil, fmt.Errorf("unable to create EventsReplayClient %w", err)
}

// Retrieve the latest emitted event
lastEventType := client.LastNEvents(ctx, 1)[0]

// Get the latest replay observable

> latestEventsObsvbl := client.EventsSequence(ctx)

Reviewer: NIT: 

```suggestion
latestEventsObs := client.EventsSequence(ctx)
```

> latestEventsObsvbl := client.EventsSequence(ctx)
> // Get the latest events from the sequence
> lastEventType := latestEventsObsvbl.Last(ctx, 1)[0]

Reviewer: ```suggestion
latestEventsObs := client.EventsSequence(ctx)
// Get the latest events from the sequence
lastEventType := latestEventsObs.Last(ctx, 1)[0]
```

    q1_conn[Connection]
    q1_dial[Dialer]
  end

  q1_obsvr1[Observer 1]
  q1_obsvr2[Observer 2]

  q1_obsvr1 --"#Subscribe()"--> q1_eb
  q1_obsvr2 --"#Subscribe()"--> q1_eb

  q1_dial =="#DialContext()"==> q1_conn
  q1_eb --"#Receive()"--> q1_conn

  q1_conn -.-> sub
```

```mermaid
---
title: EventsReplayClient Components
---

flowchart
  subgraph chain[Pocket Network]
    c1_md[Action]
  end

  subgraph erc[EventsReplayClient]
    subgraph eqc[EventsQueryClient]
      r1_qs[Event Query String]
      r1_qe[Websocket URL]
    end
    r1_rt(retryPublishEventsFactory)
    subgraph dec[EventBytesDecoder]
      r1_df[Decoder Function]
    end
    r1_ob[LatestReplayObservable]
    r1_pb[LatestReplayObservablePublisher]
  end

  subgraph con[Consumer]
    c1_ob[Replay Observable]
    c1_ev[Events]
  end

  c1_md -."#EmitTypedEvent".-> eqc

  eqc --"retry.OnError"--> r1_rt
  r1_df --"Decoded Event"--> r1_rt
  r1_rt -."#EventBytes".-> eqc
  r1_rt --"LatestReplayObservable"-->r1_pb
  r1_pb --> r1_ob

  eqc --"Serialised Event"--> r1_df
  r1_ob --"#EventsSequence"--> c1_ob
  r1_ob --"#LastNEvents"--> c1_ev
```

### Subscriptions

```mermaid
---
title: Event Subscription Data Flow

> ---

Reviewer: This diagram needs one of:

A. Its own legend, in which the usage via network I/O association direction matches the diagram
B. To be rewritten in terms of the legend provided in the previous section (i.e. turn it upside-down)
C. A note or some other indicator that the legend in the section above only applies to that section and/or does not apply to this section.

I think A is ideal but C would be sufficient for now. I think this diagram reads better as written (as opposed to upside-down). I also think that the component diagrams above read best  they way they are (i.e. shouldn't be turned upside-down either).

Author: Added todos to add legends in the appropriate places

tcpEvtClient := events.NewEventsQueryClient(cometUrl, tcpDialerOpt)

// Alternatively, with a custom gRPC dialer:
// See: https://pkg.go.dev/github.com/pokt-network/poktroll/pkg/client/events/#WithDialer
gcpDialerOpt := events.WithDialer(grcpDialer)
// See: https://pkg.go.dev/github.com/pokt-network/poktroll/pkg/client/events/#NewEventsQueryClient
grpcEvtClient := events.NewEventsQueryClient(cometUrl, grpcDialerOpt)

// ... rest follows the same as the basic example.
```

### Configuration

- **WithDialer**: Configure the client to use a custom dialer for connections.

## Usage (`EventsReplayClient`)

### Basic Usage

```go
const (
  // Define a query string to provide to the EventsQueryClient
  // See: https://docs.cosmos.network/v0.47/learn/advanced/events#subscribing-to-events
  // And: https://docs.cosmos.network/v0.47/learn/advanced/events#default-events
  eventQueryString = "message.action='eventName'"
  // Define the websocket URL the EventsQueryClient will subscribe to
  cometWebsocketURL = "ws://example.com:36657/websocket"
	// the amount of events we want before they are emitted
  replayObsBufferSize = 1
)

> // Define an interface to represent the onchain event

Reviewer: ```suggestion
// Define an interface to represent an arbitrary onchain event
```

          sub[subscribe endpoint]
      end
  end

  subgraph eqc[EventsQueryClient]
      subgraph q1[Query 1]
          q1_eb[EventsBytesObservable]
          q1_conn[Connection]
      end
      subgraph q2[Query 2]
          q2_conn[Connection]
          q2_eb[EventsBytesObservable]
      end
  end

  q1_obsvr1[Query 1 Observer 1]
  q1_obsvr2[Query 1 Observer 2]
  q2_obsvr[Query 2 Observer]

  q1_eb -.-> q1_obsvr1
  q1_eb -.-> q1_obsvr2
  q2_eb -.-> q2_obsvr

  q1_conn -.-> q1_eb
  q2_conn -.-> q2_eb

  sub -.-> q1_conn
  sub -.-> q2_conn
```

### Events Decoding

_TODO_DOCUMENTATION(@bryanchriswhite,@h5law): Add Legend_

_TODO_DOCUMENTATION(@bryanchriswhite,@h5law): separate into separate diagrams that detail each
step of the process (reconnection, decoding, publishing, etc.)_

```mermaid
---
title: Event Bytes To Type Event Sequence
---

sequenceDiagram
    participant Q as EventsQueryClient
    participant D as EventDecoder
    Q ->>+ D: EventBytes
    D ->> D: DecoderFn()
    D ->>- Q: resubscribeOnError()
    participant P as Replay Observable Cache Publish Channel
    participant O as Replay Observable Cache
    D ->> P: Typed Events
    P ->> O: Typed Events

> ```

Reviewer: Let's shelve this diagram for now (i.e. remove and iterate in [notion](https://www.notion.so/Clients-a24e75fe6f70450890c8e12092d5f8e1?d=653b16f968c34e09bed1fe06cb0e0439&pvs=4#9c938bc617374de0a95d911e1d6d2641))

> ```

Reviewer: Let's shelve this diagram for now (i.e. remove and iterate in [notion](https://www.notion.so/Clients-a24e75fe6f70450890c8e12092d5f8e1?d=653b16f968c34e09bed1fe06cb0e0439&pvs=4#9c938bc617374de0a95d911e1d6d2641)), it's not ready and isn't worth blocking this PR over.

Reviewer: :pray:
